Sources/NIOHTTP2/HTTP2Settings.swift

```diff
@@ -12,101 +12,232 @@
 //
 //===----------------------------------------------------------------------===//
 import CNIONghttp2
+import NIO
 
-/// A HTTP/2 settings parameter that allows representing both known and unknown HTTP/2
-/// settings parameters.
-public struct HTTP2SettingsParameter {
-    internal let networkRepresentation: UInt16
-
-    /// Create a `HTTP2SettingsParameter` that is not known to NIO.
+/// A single setting for HTTP/2, a combination of parameter identifier and its value.
+public enum HTTP2Setting {
```

Comment:
Let's not use an `enum` for this.

I'm already not wild about the use of an `enum` for `FramePayload`, but I don't really want to compound the mistake. A Swift `enum` is entirely non-extensible without performing a breaking release of the software. That means any new specced setting will force a brand new semver major release of this library. That's just not a good experience for anyone involved. It's better to leave this as a struct for now.

Incidentally, the struct may want to be implemented by *wrapping* a `private enum`, but that's a separate note.

Reply:
Wow. 🤦‍♂️ That's… amazingly-not-forward-thinking design of the compiler there. As an API designer by trade, Swift's enums provide a good half of the elegance of Swift for API design. Having them designed in a way that makes them essentially unusable across library boundaries is just… wow. I literally cannot describe just how frustrating and disappointing that is; I read this message just as I was going to bed last night & I couldn't get to sleep for an hour afterwards. It's like "look at all these great things! Oh, you're not working on static binaries? Yeah, you can't use these, sorry."

This basically leaves us with the only (reasonable) option being the rather convoluted model used by `ChannelOption` in SwiftNIO; I guess now I know why it has that design. Well, I know how to do that, at least, so I'll put something together using that approach, because there *has* to be something better than 'magic numbers with other magic numbers in a struct'.

Reply:
Agreed. Take a look at https://github.com/apple/swift-evolution/blob/master/proposals/0192-non-exhaustive-enums.md which aims to solve this for enums in the stdlib and overlays, but not for enums in libraries like SwiftNIO.

Reply:
Oh FFS, if they only care about fixing it for stdlib & Apple's own frameworks that's just adding insult to injury. Not to mention it severely reducing the usefulness of the Swift 5.0 stable ABI from an third-party library perspective.

Reply:
@Lukasa Given the problems with enums, would it not make more sense to define frame payloads as `protocol`/`struct` types? For example, see https://github.com/AlanQuatermain/H2Swift/blob/1e389199710ace9f6f29795892cbdd0f8dc37232/Sources/H2Swift/Frames/Frame.swift#L193 and https://github.com/AlanQuatermain/H2Swift/blob/master/Sources/H2Swift/Frames/DataFrame.swift

Reply:
…for the payload types themselves, at least.

Reply:
For frame payloads I basically propose we leave it alone for now as it's already there, and we take a separate spin at re-designing them later in the process. They should be easily-enough redesigned in isolation and it'll help focus the mind.